File: csharp/test/Vista.SDK.Tests/LocationsTests.cs
using System.ComponentModel.DataAnnotations;
using FluentAssertions;

namespace Vista.SDK.Tests;

public class LocationsTests
{
    public static IEnumerable<object[]> Test_Vis_Versions => GmodTests.Test_Vis_Versions;

    [Theory]
    [MemberData(nameof(Test_Vis_Versions))]
    public void Test_Locations_Loads(VisVersion visVersion)
    {
        var (_, vis) = VISTests.GetVis();

        var locations = vis.GetLocations(visVersion);
        Assert.NotNull(locations);
        Assert.NotNull(locations.Groups);
    }

    [Fact]
    public void Test_LocationGroups_Properties()
    {
        var values = Enum.GetValues(typeof(LocationGroup)).Cast<int>().ToArray();
        Assert.Equal(values.Length, new HashSet<int>(values).Count);
        Assert.Equal(5, values.Length); // If this changes, we need to change the way LocationCharDict is used
        Assert.Equal(0, (int)LocationGroup.Number);
        for (int i = 0; i < values.Length - 1; i++) // Start at one
        {
            Assert.Equal(i, values[i + 1] - 1);
        }
    }

    [Theory]
    [MemberData(nameof(VistaSDKTestData.AddLocationsData), MemberType = typeof(VistaSDKTestData))]
    public void Test_Locations(string value, bool success, string? output, string[] expectedErrorMessages)
    {
        var locations = VIS.Instance.GetLocations(VisVersion.v3_4a);

        var stringSuccess = locations.TryParse(value, out var stringParsedLocation, out var stringErrorBuilder);
        var spanSuccess = locations.TryParse(value.AsSpan(), out var spanParsedLocation, out var spanErrorBuilder);
        Verify(stringSuccess, stringErrorBuilder, stringParsedLocation);
        Verify(spanSuccess, spanErrorBuilder, spanParsedLocation);

        void Verify(bool succeeded, ParsingErrors errors, Location parsedLocation)
        {
            if (!success)
            {
                Assert.False(succeeded);
                Assert.Equal(default, parsedLocation);

                if (expectedErrorMessages.Length > 0)
                {
                    Assert.NotNull(errors);
                    Assert.True(errors.HasErrors);
                    var actualErrors = errors.Select(e => e.Message).ToArray();
                    actualErrors.Should().Equal(expectedErrorMessages);
                }
            }
            else
            {
                Assert.True(succeeded);
                Assert.False(errors.HasErrors);
                Assert.NotEqual(default, parsedLocation);
                Assert.Equal(output, parsedLocation.ToString());
            }
        }
    }

    [Fact]
    public void Test_Location_Parse_Throwing()
    {
        var locations = VIS.Instance.GetLocations(VisVersion.v3_4a);
        Assert.Throws<ArgumentException>(() => locations.Parse(null!));
        Assert.Throws<ArgumentException>(() => locations.Parse(ReadOnlySpan<char>.Empty));
    }

    [Fact]
    public void Test_Location_Builder()
    {
        var locations = VIS.Instance.GetLocations(VisVersion.v3_4a);

        var locationStr = "11FIPU";
        var location = locations.Parse(locationStr);

        var builder = LocationBuilder.Create(locations);

        builder = builder.WithNumber(11).WithSide('P').WithTransverse('I').WithLongitudinal('F').WithValue('U');

        Assert.Equal("11FIPU", builder.ToString());
        Assert.Equal(11, builder.Number);
        Assert.Equal('P', builder.Side);
        Assert.Equal('U', builder.Vertical);
        Assert.Equal('I', builder.Transverse);
        Assert.Equal('F', builder.Longitudinal);

        Assert.Throws<ValidationException>(() => builder = builder.WithValue('X'));
        Assert.Throws<ValidationException>(() => builder = builder.WithNumber(-1));
        Assert.Throws<ValidationException>(() => builder = builder.WithNumber(0));
        Assert.Throws<ValidationException>(() => builder = builder.WithSide('A'));
        Assert.Throws<ValidationException>(() => builder = builder.WithValue('a'));

        Assert.Equal(location, builder.Build());

        builder = LocationBuilder.Create(locations).WithLocation(builder.Build());

        Assert.Equal("11FIPU", builder.ToString());
        Assert.Equal(11, builder.Number);
        Assert.Equal('P', builder.Side);
        Assert.Equal('U', builder.Vertical);
        Assert.Equal('I', builder.Transverse);
        Assert.Equal('F', builder.Longitudinal);

        builder = builder.WithValue('S').WithValue(2);

        Assert.Equal("2FISU", builder.ToString());
        Assert.Equal(2, builder.Number);
        Assert.Equal('S', builder.Side);
        Assert.Equal('U', builder.Vertical);
        Assert.Equal('I', builder.Transverse);
        Assert.Equal('F', builder.Longitudinal);
    }

    [Fact]
    public void Test_LocationBuilder_WithLocation_SingleDigit()
    {
        var locations = VIS.Instance.GetLocations(VisVersion.v3_4a);

        // Test single digit location parsing
        var builder1 = LocationBuilder.Create(locations).WithLocation(new Location("1"));
        Assert.Equal(1, builder1.Number);
        Assert.Equal("1", builder1.ToString());

        var builder5 = LocationBuilder.Create(locations).WithLocation(new Location("5"));
        Assert.Equal(5, builder5.Number);
        Assert.Equal("5", builder5.ToString());

        var builder9 = LocationBuilder.Create(locations).WithLocation(new Location("9"));
        Assert.Equal(9, builder9.Number);
        Assert.Equal("9", builder9.ToString());

        // Test single digit with characters
        var builderMixed = LocationBuilder.Create(locations).WithLocation(new Location("1FIPU"));
        Assert.Equal(1, builderMixed.Number);
        Assert.Equal('P', builderMixed.Side);
        Assert.Equal('U', builderMixed.Vertical);
        Assert.Equal('I', builderMixed.Transverse);
        Assert.Equal('F', builderMixed.Longitudinal);
        Assert.Equal("1FIPU", builderMixed.ToString());
    }

    [Fact]
    public void Test_Locations_Equality()
    {
        var (_, vis) = VISTests.GetVis();

        var gmod = vis.GetGmod(VisVersion.v3_4a);

        var node1 = gmod["C101.663"].WithLocation("FIPU");

        var node2 = gmod["C101.663"].WithLocation("FIPU");

        Assert.Equal(node1, node2);
        Assert.NotSame(node1, node2);
    }
}

File: cpp/test/LocationsTests.cpp
#include "pch.h"

#include "dnv/vista/sdk/VIS.h"
#include "dnv/vista/sdk/Locations.h"
#include "dnv/vista/sdk/VisVersion.h"
#include "dnv/vista/sdk/ParsingErrors.h"

namespace dnv::vista::sdk::tests
{
	static std::vector<VisVersion> visVersions()
	{
		return {
			VisVersion::v3_4a,
			VisVersion::v3_5a,
			VisVersion::v3_6a,
			VisVersion::v3_7a,
			VisVersion::v3_8a };
	}

	struct LocationTestCase
	{
		std::string value;
		bool success;
		std::string output;
		std::vector<std::string> expectedErrorMessages;
	};

	class LocationsTests : public ::testing::Test
	{
	protected:
		virtual void SetUp() override
		{
			m_vis = &VIS::instance();
		}

		VIS* m_vis = nullptr;
	};

	class LocationsVersionedTests : public LocationsTests,
									public ::testing::WithParamInterface<VisVersion>
	{
	};

	TEST_P( LocationsVersionedTests, Test_Locations_Loads )
	{
		VisVersion visVersion = GetParam();

		auto locations = m_vis->locations( visVersion );

		auto groups = locations.groups();
		ASSERT_NE( groups.size(), 0 );

		Location testLocation;
		ParsingErrors errors;

		bool success = locations.tryParse( std::string_view( "11" ), testLocation, errors );
		ASSERT_TRUE( success );
		ASSERT_FALSE( errors.hasErrors() );

		ASSERT_EQ( "11", testLocation.toString() );
	}

	INSTANTIATE_TEST_SUITE_P(
		ByVisVersion,
		LocationsVersionedTests,
		::testing::ValuesIn( visVersions() ) );

	TEST_F( LocationsTests, Test_LocationGroups_Properties )
	{
		std::vector<int> values = {
			static_cast<int>( LocationGroup::Number ),
			static_cast<int>( LocationGroup::Side ),
			static_cast<int>( LocationGroup::Vertical ),
			static_cast<int>( LocationGroup::Transverse ),
			static_cast<int>( LocationGroup::Longitudinal ) };

		std::set<int> uniqueValues( values.begin(), values.end() );
		ASSERT_EQ( values.size(), uniqueValues.size() );

		ASSERT_EQ( 5, values.size() );

		ASSERT_EQ( 0, static_cast<int>( LocationGroup::Number ) );

		std::sort( values.begin(), values.end() );
		for ( size_t i = 0; i < values.size() - 1; i++ )
		{
			ASSERT_EQ( values[i] + 1, values[i + 1] );
		}
	}

	class LocationsParamTest : public LocationsTests,
							   public ::testing::WithParamInterface<LocationTestCase>
	{
	};

	void verifyParsing( bool succeeded, const ParsingErrors& errors, const Location& parsedLocation, const LocationTestCase& expected )
	{
		if ( !expected.success )
		{
			ASSERT_FALSE( succeeded );
			ASSERT_EQ( parsedLocation, Location() );

			if ( !expected.expectedErrorMessages.empty() )
			{
				ASSERT_TRUE( errors.hasErrors() );

				std::vector<std::string> actualErrors;
				for ( const auto& error : errors )
				{
					auto const& [type, message] = error;
					actualErrors.push_back( message );
				}

				ASSERT_EQ( actualErrors.size(), expected.expectedErrorMessages.size() );
				for ( const auto& expectedError : expected.expectedErrorMessages )
				{
					bool found = false;
					for ( const auto& actualError : actualErrors )
					{
						if ( actualError == expectedError )
						{
							found = true;
							break;
						}
					}
					ASSERT_TRUE( found ) << "Expected error not found: " << expectedError;
				}
			}
		}
		else
		{
			ASSERT_TRUE( succeeded );
			ASSERT_FALSE( errors.hasErrors() );
			ASSERT_NE( parsedLocation, Location() );
			ASSERT_EQ( expected.output, parsedLocation.toString() );
		}
	}

	TEST_P( LocationsParamTest, Test_Locations )
	{
		auto param = GetParam();
		auto locations = m_vis->locations( VisVersion::v3_4a );

		Location stringParsedLocation;
		ParsingErrors stringErrorBuilder;
		bool stringSuccess = locations.tryParse( std::string_view( param.value ), stringParsedLocation, stringErrorBuilder );

		Location spanParsedLocation;
		ParsingErrors spanErrorBuilder;
		bool spanSuccess = locations.tryParse( std::string_view( param.value ), spanParsedLocation, spanErrorBuilder );

		verifyParsing( stringSuccess, stringErrorBuilder, stringParsedLocation, param );
		verifyParsing( spanSuccess, spanErrorBuilder, spanParsedLocation, param );
	}

	std::vector<LocationTestCase> locationsData()
	{
		return {
			{ "11FIPU", true, "11FIPU", {} },
			{ "11FI", true, "11FI", {} },
			{ "11F", true, "11F", {} },
			{ "11", true, "11", {} },
			{ "F", true, "F", {} },
			{ "FIPU", true, "FIPU", {} },
			{ "XYZ", false, "", { "Invalid location code: 'XYZ' with invalid location code(s): 'X','Y','Z'" } },
		};
	}

	INSTANTIATE_TEST_SUITE_P(
		LocationParseData,
		LocationsParamTest,
		::testing::ValuesIn( locationsData() ) );

	TEST_F( LocationsTests, Test_Location_Parse_Throwing )
	{
		auto locations = m_vis->locations( VisVersion::v3_4a );

		ASSERT_THROW( { (void)locations.parse( std::string_view() ); }, std::invalid_argument );
		ASSERT_THROW( { (void)locations.parse( std::string_view( "" ) ); }, std::invalid_argument );
		ASSERT_THROW( { (void)locations.parse( std::string_view( nullptr, 0 ) ); }, std::invalid_argument );
	}

	TEST_F( LocationsTests, Test_Location_Builder )
	{
		// TODO Implement this test when LocationBuilder is available
	}

	TEST_F( LocationsTests, Test_Locations_Equality )
	{
		const auto& gmod = m_vis->gmod( VisVersion::v3_4a );

		auto node1 = gmod["C101.663"].withLocation( "FIPU" );
		auto node2 = gmod["C101.663"].withLocation( "FIPU" );

		ASSERT_EQ( node1, node2 );
		ASSERT_NE( &node1, &node2 );
	}
}
